### ISSUE TITLE
Do the opposite of `let_and_return`

### ISSUE THREAD

**Issue by amab8901** [rust-lang/rust-clippy#12512]
### What it does

The opposite of [let_and_return](https://rust-lang.github.io/rust-clippy/master/index.html#/let_and_return). It will find which parts of your code are directly instantiating something as an output return value without first assigning it to a variable, and it will tell you to first assign to a variable before returning the value.  

### Advantage

The `let_and_return` lint's ["Why is this bad?"](https://rust-lang.github.io/rust-clippy/master/index.html#/let_and_return) explanation says the following:
```
It is just extraneous code. Remove it to make your code more rusty.
```
I strongly disagree with both of these sentences. 

**"extraneous code"**
The word "extraneous" suggests that this extra code doesn't serve any purpose and has no value. But that code does indeed serve a purpose and bring value. Here are some reasons for using the let_and_return pattern:

* If you're debugging the code, you might want to add `dbg!()` or `println!()` statement on the return value to see what value is being returned. 
* A new contributor to a repository wants to explore the repository to understand how it works. `dbg!()` and `println!()` wrapper around the return value helps the new contributor understand how the data values get changed by each line of code, including the last line. 
* It's more convenient to put a wrapper around a single variable name in the end instead of having to put a dbg!() or println!() wrapper around a large unit of code (for example a struct that gets instantiated directly into the last line as a return value).
* The variable name chosen for the return value serves as a form of documentation since it explains to the one exploring the repository how this data should be interpreted and what purpose it serves in the broader context. 
* The let_and_return pattern doesn't incur any performance penalties since the compiler will optimize it away during the compilation process. 
* By doing as I suggested above (in "What it does" section), we can avoid/prevent the "Known Problem" that the current `let_and_return` lint admits to causing - namely deadlocks as elaborated in https://github.com/rust-lang/rust/issues/37612. 
```

### Drawbacks

* Getting fewer points in code golf where the goal is to implement a given algorithm as quickly as possible and with as few words as possible
* A few extra words to read in the source code
* Upfront work to implement this lint for Clippy maintainers and for end-user to apply this lint.

### Example

```rust
fn foo() -> String {
    String::new()
}
```

Could be written as:

```rust
fn foo() -> String {
    let x = String::new();
    x
}
```

**Comment by itsfarseen** [comment]
`let_and_return` is the most annoying lint in clippy. 
Often I have code like this: 

```rust
let semantic_name = complex calculation;
semantic_name
```

This is a very useful thing to have. Arguably, the function name, or the enclosing block should clearly convey the semantic meaning of the calculation. But often times, having a new label at the calculation site makes it much easier to read.


For example:
```rust
let salary_by_month = {
  if month == "feb" {
     let salary_in_feb = calculate_salary(28)
     salary_in_feb
  } else {
     let normal_salary = calculate_salary(30)
     normal_salary
  }
}
```

**Comment by amab8901** [comment]
yeah especially problematic with match statements and conditionals where different return values can have different indentations so you might accidentally miss one of them when debugging (or need to take longer to find them all, or put separate debug statements on each match arm or each conditional block). `let_and_return` lets you bundle all those into one variable and put dbg!() on that one variable instead of having to go through every conditional and every match arm manually.

**Comment by TomFryersMidsummer** [comment]
What makes
```rust
    pub fn new(x: f32, y: f32) -> Self {
        dbg!(Self { x, y })
    }
```
more difficult to write than
```rust
    pub fn new(x: f32, y: f32) -> Self {
        let new = Self { x, y };
        dbg!(new)
    }
```
?

Both are just adding six characters (or pressing `C-c` `C-d`, or whatever your editor's shortcut is, once).

**Comment by itsfarseen** [comment]
I think @amab8901 meant this:

```rust
fn classify(n: i32) -> &'static str {
    let result = match n {
        x if x < 0 => "negative",
        0 => "zero",
        1..=9 => "small",
        _ if n % 2 == 0 => "even",
        _ => "odd",
    };

    // Handy: one debug point for *all* match arms
    dbg!(result);

    result
}
```

this is much easier to do compared to the below example because the closing parenthesis is 2 miles away.

```rust
fn classify(n: i32) -> &'static str {
    dbg!(match n {
        x if x < 0 => "negative",
        0 => "zero",
        1..=9 => "small",
        _ if n % 2 == 0 => "even",
        _ => "odd",
    })
}
```

or worse

```rust
fn classify(n: i32) -> &'static str {
    match n {
        x if x < 0 => dbg!("negative"),
        0 => dbg!("zero"),
        1..=9 => dbg!("small"),
        _ if n % 2 == 0 => "even", // oops, missed this
        _ => dbg!("odd"),
    }
}
```

**Comment by amab8901** [comment]
yeah kinda like @itsfarseen demonstrated. The issue isn't which one is easier to write. The issue is which one is easier to troubleshoot when there's a bug in a complicated program and you have no idea where the bug is coming from. 

# with let_and_return pattern
If we use the `let_and_return` pattern, then we can just double-click `result`, ctrl-C, and paste it inside the parentheses of `dbg!()`. Once we're done, we can just select that line and delete it.


# without let_and_return pattern:
But if we use the pattern currently recommended by Clippy and if we have a large function with a match statement that's 90 lines long, we will have to scroll to the top of the match statement and click the start of the match statement (and be careful not to click on the wrong character) and write `dbg!(`, and then the `rust-analyzer` will make the code annoyingly noisy red/yellow color all over the place. And with this noisy red/yellow screen I now have to scroll down to the bottom of the 90-lines long match statement and click the end of it and write `);`. And when I'm done debugging, I have to go through the same relatively tedious process of scrolling to the bottom of the 90-lines long match statement and remove `);` and get this yellow/red all over the place while I scroll up to find where the start is, then click there and remove the `dbg!(`;

**Comment by TomFryersMidsummer** [comment]
Even assuming you haven't set up a shortcut (which, if you're using `dbg!` so often that it's dictating code style, you ought to), your first `result` example requires, in Vim,

<kbd>i</kbd> <kbd>d</kbd> <kbd>b</kbd> <kbd>g</kbd> <kbd>!</kbd> <kbd>(</kbd> <kbd>ESC</kbd> <kbd>$</kbd> <kbd>a</kbd> <kbd>)</kbd> <kbd>ESC</kbd>

whereas the second version without `result` needs

<kbd>i</kbd> <kbd>d</kbd> <kbd>b</kbd> <kbd>g</kbd> <kbd>!</kbd> <kbd>(</kbd> <kbd>ESC</kbd> <kbd>$</kbd> <kbd>%</kbd> <kbd>a</kbd> <kbd>)</kbd> <kbd>ESC</kbd>

(The third version with `dbg!` on every arm is just silly.)

That's one extra key-stroke. Deleting it again is similar. I don't see how saving one key-stroke is enough benefit to be worth adding a redundant line of code to every function.

If your text editor makes this process significantly more difficult than this, then that's a problem with the text editor.

**Comment by y21** [comment]
I agree with `@TomFryersMidsummer` that there are other ways to solve the `dbg!` ergonomics issue, if it actually is one, before putting in the work to implement a lint for this and *enforcing* a trailing `let` in every single function.

To add another very common editor aside from vim, I use VSCode, and it's still very easy to add a `dbg!` to the return value regardless of how many lines it makes up; simply add `.dbg` after the `}` of the trailing match, hit ctrl-space to get the snippet suggestion and hit enter.

------

With closures and iterator chains this lint feels like it would become very noisy. An iterator chain like
```rs
    let output = input.iter().filter(|v| v > 3).map(|v| v * 3).sum();
```
would presumably now have to be written...
```rs
    let output = input
        .iter()
        .filter(|v| {
            let out = v > 3;
            out
        })
        .map(|v| {
            let out = v * 3;
            out
        })
        .sum();
```

> The variable name chosen for the return value serves as a form of documentation since it explains to the one exploring the repository how this data should be interpreted and what purpose it serves in the broader context.

This has been answered before but I think that in the vast majority of cases, it feels like the variable name would not add much value (be something like "output" or "result" or something like that), whereas the function name itself already conveys the meaning.


That's not to say that there *aren't* cases where it makes sense to give the return value a different semantic name and `let_and_return` isn't helpful. But I don't think a new lint for enforcing the opposite is the solution.

**Comment by itsfarseen** [comment]
Ah, I misunderstood. This is an issue for creating a new lint. I just wanted to advocate for abolishing the `let_and_return` lint.
